crates/core/tedge_api/src/address.rs

```diff
@@ -130,6 +130,15 @@ impl<M: Message> ReplySender<M> {
             .send(Box::new(msg))
             .map_err(|msg| *msg.downcast::<M>().unwrap())
     }
+
+    /// Check whether the ReplySender is closed
+    ///
+    /// This function returns when the internal communication channel is closed.
+    /// This can be used (with e.g. [tokio::select]) to check whether the message sender stopped
+    /// waiting for a reply.
+    pub async fn closed(&mut self) {
+        self.reply_sender.closed().await
+    }
```

Comment:
As already said, I would prefer to have messages without a reply built-in. Having this option for the sender to tell its peer that no response is expected meets the need but seems contrived. I would prefer the reverse: the sender telling its peer that a response is expected and a channel prepared for that.

Reply:
I am not sure why this would be easier to use, since a message and its reply are hardcoded in the `Message` implementation. So I would assume that it will be mainly added for messages that actually expect a reply. For everything else there would be some variation of `NoReply`, which is effectively a ZST and thus has a minimal impact overall.

I am sure that this is something that can be changed in a later PR where we can discuss working alternatives.

Reply:
I have a mix feeling when I see that a `Message` type has an associated `Reply` type. Most messages expect no response (e.g. telemetry data). Okay we can use this zero-sized type `NoReply`. But there is a complexity overhead to establish the one-shot channel. Furthermore, some messages expects several responses (e.g. Cumulocity expects an executing status then the outcome for each operation request). I guess we can find a way to deal with that, but for me this `Reply` adds more complexity than values.

My main concern is the mix of communication patterns. A plugin receives messages from two different channels. The requests are queued in a channel associated to their address. The responses are send over one-shot channels. The plugin will have to select between these channels: process the next request or wait for a response.

I would prefer a *single* channel where a plugin receives *both* inputs and responses to its own requests to other plugins. I agree that the counterpart of this approach is that a plugin might have to handle a state machine to keep track of the pending responses. While with this PR, as far I understand it, a plugin can spawn cooperative threads instead.

Reply:
> But there is a complexity overhead to establish the one-shot channel. 

Not sure what complexity you mean here. A plugin author would not have to deal with this at all. The core implementation is responsible for this! If you mean complexity as in runtime cost: This is absolutely negligible, we're talking about a few bytes + the size of the reply message type.

> Furthermore, some messages expects several responses (e.g. Cumulocity expects an executing status then the outcome for each operation request). I guess we can find a way to deal with that, but for me this Reply adds more complexity than values.

Right now, having several replies is as easy as using a `Vec` as response type! (`struct MyReply(Vec<_>);`). 
But maybe I am missing some context here. Could you provide an example of how such a "request->multi response" pattern would look like (if possible with some pseudo-code)?

If we would remove replies, how would a plugin send these? What I could think of would be a custom message types that the receiving plugin needs to handle, which would mean a `Handle` implementation for each possible reply. Plus some internal state machinery would be required to make sure the plugin is in the correct state when a reply is received.
To me this sounds like a lot complexity just for receiving replies!

> The plugin will have to select between these channels: process the next request or wait for a response.

No, there is no selecting involved. The `Handle` implementation takes care of processing the next request. If inside a request handler, a plugin has to wait for a response. There's no additional complexity involved!

> I agree that the counterpart of this approach is that a plugin might have to handle a state machine to keep track of the pending responses. While with this PR, as far I understand it, a plugin can spawn cooperative threads instead.

Cooperative threads are not needed for handling messages or replies. These are only required if there's communication from "the outside world" to thin-edge, for example if the plugin is an ingest for MQTT.

Reply:
> > Furthermore, some messages expects several responses (e.g. Cumulocity expects an executing status then the outcome for each operation request). I guess we can find a way to deal with that, but for me this Reply adds more complexity than values.
> 
> Right now, having several replies is as easy as using a `Vec` as response type! (`struct MyReply(Vec<_>);`). But maybe I am missing some context here. Could you provide an example of how such a "request->multi response" pattern would look like (if possible with some pseudo-code)?

I mean several *independent* messages not a bunch of messages. For instance, sending progress status for an operation request, `Scheduled`, `Started` then `Done`.

Reply:
Trying again to explain why I have a mix feeling with this `Reply` type.

Let's put aside the runtime & memory overhead. I agree this is currently not a key point.

My issue is more about the mental model behind these plugins. What is the general idea? The main concept ? The underlying design pattern?

I understand what is proposed here through the lenses of actors: "stand-alone threads that manage a state, react to a stream of input messages, produces a stream of output messages and interact with each others only using these streams of messages". I expect to see *all* the inputs of an actor/plugin going through its input channel, *including* the responses to its requests. I expect to see *all* the outputs of an actor/plugin through its output channel, *including* responses sent to a specific requester. I see that as a foundation for observability, debug traceability and testability.

What I see here is an hybrid model. Some messages are queued and processed one after the other by a plugin. Some messages are sent along a direct channel and processed with a different priority. If I want to implement something as the tedge mapper - that process telemetry data and operations, I will have to find a way to wait for the final outcome of the operations without blocking telemetry data. I can imagine solutions to that, but I currently don't understand how the current proposal help me.

Reply:
> What is the general idea?

The general idea is to have a direct chain of messages between Plugins. Each plugin gets to choose who they send messages to (through the `Address` struct) and 1. if they want to wait for a reply, 2. how long (if at all).

```
                    ┌────────────────────┐  ┌────────────────────┐
                    │    Plugin A        │  │    Plugin B        │
                    │                    │  │                    │
                    └─────────┬──────────┘  └──────────┬─────────┘
                              │                        │
┌─────────────┐               │                        │
│   Message   ├────────┐      │                        │
└─────────────┘        │      │                        │
                       │      │                        │
                       │     ┌┴─┐                      │
                       └─────►  │                      │
                             │  ├─────────────┐        │
                             │  │             │        │
                             │  │             │     ┌──┴─┐
                             │  │             └─────►    │
                             │  │                   │    │
                             │  │                   │    │
                             │  │       ┌──?────────┤    │
                             │  │       │           └──┬─┘
                             │  │       │              │
                             │  ◄───────┘              │
                             │  │                      │
                             └┬─┘                      │
                              │                        │
                              │                        │
```


This has the following benefits:

1. Plugin authors are able to represent a state machine through the `Handle<M>` implementation as they can move in lockstep with replies. You mentioned for example a list of work to be done and waiting for feedback. In this situation that can be done easily (either sequentially or concurrently) and the plugin is easy to write. If one were to enforce that all replies come in through the same channel as 'normal' messages, one would have to somehow tag/recognize them and also handle them correctly in the original sending context. (This also raises questions about how long to keep this context, since replies are not guaranteed for a variety of reasons)
1. All messages go through API-controlled interfaces, this means that not only can everything be traced, they can also be constructed by users to test their own Plugins. (For example there is an `Address::new`, but this needs to be potentially expanded which is not an issue)
1. Since the `tracing` library allows for tracing whole `Future`s having this back-and-forth between plugins naturally shows the lifetime of a single message handler. 


> What I see here is an hybrid model.

Where exactly is _here_ ? Messages sent through `Address::send*` are only ever handled by a `Handle<M>` implementation. This implementation can also send messages through either the backchannel to the requestor _or_ through their own `Address` structs.

> Some messages are queued and processed one after the other by a plugin.

All messages sent trough `Address::send` are queued and then processed concurrently by the plugin's relevant `Handle<M>` implementation.

> Some messages are sent along a direct channel and processed with a different priority.

All channels are direct and only thinly wrapped for ease-of-use. No prioritization happens anywhere.

> If I want to implement something as the tedge mapper - that process telemetry data and operations, I will have to find a way to wait for the final outcome of the operations without blocking telemetry data.

You can simply `.await` the operation, all other messages are handled concurrently.

> I can imagine solutions to that, but I currently don't understand how the current proposal help me.

The current proposal already handles this situation.